cross-media-measurement/src/main/kotlin/org/wfanet/measurement/duchy/daemon/mill/liquidlegionsv2/LiquidLegionsV2Mill.kt

```diff
@@ -516,12 +525,51 @@ class LiquidLegionsV2Mill(
       )
   }
 
+  // Combines the public keys from the duchies after this duchy in the ring and the primary duchy.
+  fun getPartiallyCombinedPublicKey(globalId: String, nextDuchy: String): ElGamalPublicKey {
+    val order = duchyOrder.computationOrder(globalId)
+    require(order.contains(nextDuchy)) { "$nextDuchy doesn't exist in the duchy ring." }
+    if (nextDuchy == order[0]) {
+      // Return the aggregator's key if next duchy is the aggregator.
+      return cryptoKeySet.otherDuchyPublicKeys[nextDuchy] ?: error(
+        "$nextDuchy is not in the key set."
+      )
+    }
+
+    val partialList = order.subList(order.indexOf(nextDuchy), order.size) + order[0]
+    val request = CombineElGamalPublicKeysRequest.newBuilder().apply {
+      curveId = cryptoKeySet.curveId.toLong()
+      addAllElGamalKeys(
+        partialList.map {
+          val key = cryptoKeySet.otherDuchyPublicKeys[it] ?: error(
+            "$it is not in the key set."
+          )
+          ElGamalPublicKeys.newBuilder().apply {
+            elGamalG = key.generator
+            elGamalY = key.element
+          }.build()
+        }
+      )
+    }.build()
+    val response = CombineElGamalPublicKeysResponse.parseFrom(
```

Comment:
It could be out of scope for this PR but should we cache these results?

Reply:
the look up is pretty cheap comparing to what the main protocol would do. 
but yep, caching the result is still good. This could be stored in a in memory map in the mill.
Done